jupyterhub_chameleon/handler.py

```diff
@@ -184,76 +137,116 @@ async def _fetch_new_token(self, refresh_token):
+            if not artifact_id:
+                raise HTTPError(
+                    requests.codes.bad_request, "Could not understand import request"
+                )
 
+            if not self.current_user:
+                raise HTTPError(
+                    requests.codes.unauthorized,
+                    "Authentication with API token required",
+                )
 
-class ArtifactPublishPrepareUploadHandler(AccessTokenMixin, APIHandler):
-    async def get(self):
-        if not self.current_user:
-            raise HTTPError(401, "Authentication with API token required")
+            await self.refresh_token(source="artifact_import_redirect")
+            auth_state = await self.current_user.get_auth_state()
 
-        # For federated logins, Keystone authentication relies on short-lived
-        # tokens from the IdP; ensure we have a fresh one stored for the user.
-        await self.refresh_token(source="artifact_publish_prepare")
+            trovi_token = trovi.exchange_token(auth_state.get("access_token"))
 
-        auth_state = await self.current_user.get_auth_state()
+            artifact = trovi.fetch_artifact(artifact_id, trovi_token)
 
-        # Authenticate to Trovi with our OpenStack credentials,
-        # exchanging our keystone token for a Trovi token
-        trovi_resp = requests.post(
-            urljoin(os.getenv("TROVI_URL"), "/token/"),
-            headers={"Content-Type": "application/json", "Accept": "application/json"},
-            json={
-                "grant_type": "token_exchange",
-                "subject_token": auth_state.get("access_token"),
-                "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
-                "scope": "artifacts:read artifacts:write",
-            },
-        )
+            versions = list(
+                sorted(
+                    artifact["versions"],
+                    reverse=True,
+                    key=lambda v: v["created_at"],
+                )
+            )
+            if version_number < 0 or version_number >= len(versions):
+                raise HTTPError(requests.codes.bad_request, "Invalid version number")
+            if len(versions) == 0:
+                raise HTTPError(
+                    requests.codes.internal_server_error,
+                    "Artifact has no linked content.",
+                )
+            version = versions[version_number]
 
-        trovi_token = trovi_resp.json()
+            contents_url = trovi.fetch_contents_url(
+                version["contents"]["urn"], trovi_token
+            )
 
-        self.log.debug(auth_state.get("access_token"))
+            owner = artifact["owner_urn"].split(":")[-1]
+            if self.current_user.name == owner or not artifact.id:
+                self.ownership = "own"
+            else:
+                self.ownership = "fork"
 
-        self.log.debug(f"Trovi token exchange response: {trovi_token}")
+            self.artifact = artifact
```

Comment:
I don't understand how this can work? It's storing the artifact, but then immediately redirecting, which ends the request and tears down the handler. So I think this reference will be lost?